mithril-stm/src/stm.rs

```diff
@@ -879,6 +973,69 @@ impl<D: Clone + Digest + FixedOutput + Send + Sync> StmAggrSig<D> {
     }
 }
 
+impl<D: Digest + FixedOutput> FullNodeVerifier<D> {
+    /// Verification for a full node verifier
+    pub fn verify(
+        &self,
+        signatures: &[StmSig],
+        parameters: &StmParameters,
+        msg: &[u8],
+    ) -> Result<(), StmAggregateSignatureError<D>> {
+        let signed_parties = self.collect_signed_parties(signatures);
+        let mut nr_indices = 0;
+        let mut unique_indices = HashSet::new();
+
+        for (i, sig) in signatures.iter().enumerate() {
+            sig.check_indices(parameters, &signed_parties[i].1, msg, &self.total_stake)?;
+            for &index in &sig.indexes {
+                unique_indices.insert(index);
+                nr_indices += 1;
+            }
+        }
+
+        if nr_indices != unique_indices.len() {
+            return Err(StmAggregateSignatureError::IndexNotUnique);
+        }
+
+        if (nr_indices as u64) < parameters.k {
+            return Err(StmAggregateSignatureError::NoQuorum);
+        }
+
+        let signatures = signatures
+            .iter()
+            .map(|sig| sig.sigma)
+            .collect::<Vec<Signature>>();
+        let vks = signed_parties
+            .iter()
+            .map(|reg_party| reg_party.0)
+            .collect::<Vec<VerificationKey>>();
+
+        Signature::verify_aggregate(msg.to_vec().as_slice(), &vks, &signatures)?;
+
+        Ok(())
```

Comment:
This looks very similar to `preliminary_verify` and `verify` from `StmAggrSig`. We should try to put them together, and minimise code duplication. Any reason why this would not be possible?